Fix Gradle 7.1 deprecation warnings

boris-petrov opened the pull request:
Fixes #49 

cc @rwinch

Not sure how to test this as any Gradle command that I run leads to:

```
No such property: COMPILE_CONFIGURATION_NAME for class: org.gradle.api.plugins.JavaPlugin
```

Any ideas?

rwinch commented:
Thanks for the PR @boris-petrov!

Did you run using the existing Gradle Wrapper instead of gradle 7.1?

For this PR, can you also please be sure to add a test? See https://github.com/spring-io/nohttp/blob/0.0.8/nohttp-gradle/src/test/kotlin/io/spring/nohttp/gradle/NoHttpCheckstylePluginITest.kt#L53 and https://github.com/spring-io/nohttp/blob/0.0.8/nohttp-gradle/src/test/kotlin/io/spring/nohttp/gradle/NoHttpCheckstylePluginITest.kt#L175

boris-petrov commented:
@rwinch - thanks for the tip! I always forget that Gradle projects have a local wrapper... :)

I removed Gradle 7.0.1 from testing and added 7.1 instead. Also, I fixed the code as it was not working before.

Unfortunately I still can't run the tests. I get:

```
FATAL ERROR in native method: processing of -javaagent failed, processJavaStart failed
*** java.lang.instrument ASSERTION FAILED ***: "result" with message agent load/premain call failed at src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 422
Process 'Gradle Test Executor 12' finished with non-zero exit value 134
org.gradle.process.internal.ExecException: Process 'Gradle Test Executor 12' finished with non-zero exit value 134
        ...
Caused by: java.lang.NoSuchFieldException: $jacocoAccess
        at java.base/java.lang.Class.getField(Class.java:2092)
        at org.jacoco.agent.rt.internal_28bab1d.core.runtime.ModifiedSystemClassRuntime.createFor(ModifiedSystemClassRuntime.java:138)
        ... 9 more
```

I'm running on JDK 15, that might be the issue, not sure. More ideas? :smiley:

rwinch commented:
Please just add 7.1 and do not remove 7.0.1 which was added to test gh-47.

The project builds with JDK 8 so I would try that. If it doesn't work, please post the command you entered and the complete error message.

boris-petrov commented:
OK, I returned Gradle 7.0.1 to the test matrix.

As for the tests:

```sh
./gradlew test -Dorg.gradle.java.home=/usr/lib/jvm/java-8-jdk
```

Running that on my PR seems to fail all tests on 6.0.1:

```
io.spring.nohttp.gradle.NoHttpCheckstylePluginITest > usesDefaultConfigDirectory[Gradle 6.0.1] FAILED
    org.gradle.testkit.runner.UnexpectedBuildFailure at NoHttpCheckstylePluginITest.kt:133

io.spring.nohttp.gradle.NoHttpCheckstylePluginITest > worksIfCheckstyleConfigIsPresentInDefaultLocation[Gradle 6.0.1] FAILED
    org.gradle.testkit.runner.UnexpectedBuildFailure at NoHttpCheckstylePluginITest.kt:156

io.spring.nohttp.gradle.NoHttpCheckstylePluginITest > fromCache[Gradle 6.0.1] FAILED
    org.gradle.testkit.runner.UnexpectedBuildFailure at NoHttpCheckstylePluginITest.kt:100

io.spring.nohttp.gradle.NoHttpCheckstylePluginITest > httpIsFailed[Gradle 6.0.1] FAILED
    java.lang.AssertionError at NoHttpCheckstylePluginITest.kt:78

io.spring.nohttp.gradle.NoHttpCheckstylePluginITest > httpsIsSuccess[Gradle 6.0.1] FAILED
    org.gradle.testkit.runner.UnexpectedBuildFailure at NoHttpCheckstylePluginITest.kt:66

io.spring.nohttp.gradle.NoHttpCheckstylePluginITest > upToDate[Gradle 6.0.1] FAILED
    org.gradle.testkit.runner.UnexpectedBuildFailure at NoHttpCheckstylePluginITest.kt:88

io.spring.nohttp.gradle.NoHttpCheckstylePluginITest > fromCacheWhenSwitchDirectories[Gradle 6.0.1] FAILED
    org.gradle.testkit.runner.UnexpectedBuildFailure at NoHttpCheckstylePluginITest.kt:114

58 tests completed, 7 failed

> Task :nohttp-gradle:test FAILED

FAILURE: Build failed with an exception.
```

The failures are of the likes of:

```
java.lang.NoSuchMethodError: org.gradle.api.tasks.JavaExec.getMainClass()Lorg/gradle/api/provider/Property;
```

I guess these are because the new properties that I use were not implemented at that point.

I guess one solution is to remove Gradle 6.0.1 (which is very old now) from the test matrix.

rwinch commented:
I don't plan on removing support for older versions unless we absolutely have to. The support likely needs to be added using conditional behavior based on the Gradle version. For example: https://github.com/spring-io/nohttp/commit/6fb1b25b770a2b0db77731db46d13a9bff914b0e It may be required to use reflection to support this.

boris-petrov commented:
I don't think that supporting some horrible reflective code is worth for supporting a Gradle version which is almost two years old. Upgrading from Gradle 6.0 to Gradle 6.8 (where this code works fine) is a no-brainer. It probably works on much older Gradle versions than 6.8 too.

I could try doing it but as I said - I think it's a bad idea.

rwinch commented:
Thank you for your perspective. I do not plan on removing support for Gradle 6.0 at this time.

rwinch commented:
Thanks for the Pull Request! This is now merged into main :smile: This is now released in 0.0.10

boris-petrov commented:
@rwinch - thanks for the merge and release! However, I believe you've forgotten to publish `nohttp-checkstyle:0.0.10`:

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'data-server'.
> Could not resolve all artifacts for configuration ':classpath'.
   > Could not find io.spring.nohttp:nohttp-checkstyle:0.0.10.
     Searched in the following locations:
       - https://plugins.gradle.org/m2/io/spring/nohttp/nohttp-checkstyle/0.0.10/nohttp-checkstyle-0.0.10.pom
     If the artifact you are trying to retrieve can be found in the repository but without metadata in 'Maven POM' format, you need to adjust the 'metadataSources { ... }' of the repository declaration.
     Required by:
         project : > io.spring.nohttp:io.spring.nohttp.gradle.plugin:0.0.10 > io.spring.nohttp:nohttp-gradle:0.0.10
```

rwinch commented:
@boris-petrov Sorry for the trouble. There were some issues with the artifacts getting to Maven Central. The Can you please try again?

boris-petrov commented:
@rwinch no worries! All is fine now! Thanks again!
